#  Host-gut microbiota interactions shape parasite infections in farmed Atlantic salmon
Full record of the statistical analyses performed for Brealey *et al.* (2023) Host-gut microbiota interactions shape parasite infections in farmed Atlantic salmon. Currently available as a preprint on *bioRxiv*: https://doi.org/10.1101/2023.07.20.549827.

## Abstract
Animals and their associated microbiota share long evolutionary histories. However, it is not always clear how host genotype and microbiota interact to affect phenotype. We applied a hologenomic approach to explore how host-microbiota interactions shape lifetime growth and parasite infection in farmed Atlantic salmon. Multi-omics datasets were generated from the guts of 460 salmon, 82% of which were naturally infected with an intestinal cestode. One salmonid-specific *Mycoplasma* dominated the gut metagenome of uninfected salmon. However, the microbiota was perturbed in smaller, parasitised fish, with increased abundance of *Vibrionaceae* and other *Mycoplasma* species previously linked to the cestode microbiota. The cestode-associated *Mycoplasma* carried more virulence-associated genes than the salmonid *Mycoplasma*. Colonisation by one cestode-associated *Mycoplasma* was associated with genomic variation in salmon long noncoding RNAs previously implicated in host control of intestinal microbiota. Integrating the multi-omic datasets revealed coordinated changes in the salmon gut transcriptome and metabolome that correlated with shifts in the microbiota of smaller, parasitised fish. Our results suggest that cestode infections trigger host responses and that the hostâ€™s genetic background can benefit the establishment of potentially pathogenic microbes in the gut. Our study highlights the value of a hologenomic approach for gaining an in-depth understanding of trilateral interactions among host, microbiota and parasite.

## Files and data
All statistical analyses can be viewed in the .html file, with details in the associated R markdown (.Rmd) script. The processed data files required to run the script are available in data/. Outputs from specific analyses are available in the output_*/ directories. Figures and tables (main and supplementary) are available in figures/ and tables/, respectively.

Raw nucleotide sequencing data will soon be available on the [European Nucleotide archive](https://www.ebi.ac.uk/ena/browser/home) under project PRJEB64334. Raw metabolomic data will be available on [MassIVE](https://massive.ucsd.edu/ProteoSAFe/static/massive.jsp) under the dataset identifier XXXX (TBD).
